<label style='color:red'> ¡Completar esta celda! </label>

__Integrantes:__

+ Nombre: , Grupo (día):
+ Nombre: , Grupo (día):

# Unidad 1

**Ejercicio 1.**

Instale la libreria [*NLTK*](http://www.nltk.org/howto/twitter.html) e imprima la versión de esta

In [10]:
import nltk as nltk
import numpy as np

In [12]:
nltk.__version__

'3.7'

**Ejercicio 2.**

Ejecute el las siguientes lineas de código e imprima por pantalla los primeros 100 registros del *train_x*. ¿Presenta algún error? si es así escriba la solución en un comentario 

```python

from nltk.corpus import twitter_samples 

all_positive_tweets = twitter_samples.strings('positive_tweets.json')
all_negative_tweets = twitter_samples.strings('negative_tweets.json')

tweets = all_positive_tweets + all_negative_tweets ## Concatenate the lists. 
labels = np.append(np.ones((len(all_positive_tweets),1)), np.zeros((len(all_negative_tweets),1)), axis = 0)

train_pos  = all_positive_tweets[:4000]
train_neg  = all_negative_tweets[:4000]

train_x = train_pos + train_neg 
```

In [8]:
#Presentamos un error
#Solución, usando download
#import nltk
#nltk.download('twitter_samples')
#twitter_samples.fileids()

In [13]:
nltk.download('twitter_samples')

all_positive_tweets = twitter_samples.strings('positive_tweets.json')
all_negative_tweets = twitter_samples.strings('negative_tweets.json')

tweets = all_positive_tweets + all_negative_tweets ## Concatenate the lists. 
labels = np.append(np.ones((len(all_positive_tweets),1)), np.zeros((len(all_negative_tweets),1)), axis = 0)

train_pos  = all_positive_tweets[:4000]
train_neg  = all_negative_tweets[:4000]

train_x = train_pos + train_neg 

[nltk_data] Downloading package twitter_samples to C:\Users\Juan
[nltk_data]     Pablo\AppData\Roaming\nltk_data...
[nltk_data]   Package twitter_samples is already up-to-date!


In [16]:
print(train_x[0:100])

['#FollowFriday @France_Inte @PKuchly57 @Milipol_Paris for being top engaged members in my community this week :)', '@Lamb2ja Hey James! How odd :/ Please call our Contact Centre on 02392441234 and we will be able to assist you :) Many thanks!', '@DespiteOfficial we had a listen last night :) As You Bleed is an amazing track. When are you in Scotland?!', '@97sides CONGRATS :)', 'yeaaaah yippppy!!!  my accnt verified rqst has succeed got a blue tick mark on my fb profile :) in 15 days', '@BhaktisBanter @PallaviRuhail This one is irresistible :)\n#FlipkartFashionFriday http://t.co/EbZ0L2VENM', "We don't like to keep our lovely customers waiting for long! We hope you enjoy! Happy Friday! - LWWF :) https://t.co/smyYriipxI", '@Impatientraider On second thought, there’s just not enough time for a DD :) But new shorts entering system. Sheep must be buying.', 'Jgh , but we have to go to Bayan :D bye', 'As an act of mischievousness, am calling the ETL layer of our in-house warehousing app Katam

**Ejercicio 3.**

Ejecute las siguientes lineas de código
```python
import re 

tweet = all_positive_tweets[2277]

print('\033[92m' + tweet)
print('\033[94m')

tweet2 = re.sub(r'https?:\/\/.*[\r\n]*', '', tweet)

tweet2 = re.sub(r'#', '', tweet2)

print(tweet2)
```
+ Investigue el proposito de la libreria `re` y escriba este resultado en una celda markdown
+ Describa en una celda markdown el resultado de la ejecución de cada `re.sub`
+ Convierta *tweet2* a minusculas.

In [23]:
import re 

tweet = all_positive_tweets[2277]

print('\033[92m' + tweet)
print('\033[94m')

tweet2 = re.sub(r'https?:\/\/.*[\r\n]*', '', tweet)
tweet2 = re.sub(r'#', '', tweet2)

print(tweet2)

My beautiful sunflowers on a sunny Friday morning off :) #sunflowers #favourites #happy #Friday off… https://t.co/3tfYom0N1i

My beautiful sunflowers on a sunny Friday morning off :) sunflowers favourites happy Friday off… 


**RegEx Python**
Regular expression, es una secuencia de caracteres que forman un patrón de búsqueda
Puede ser utilizada para revisar si una cadena contiene el patrón especificado.

Tiene las siguientes funciones:
+ findall(): Retorna una lista con todas las coincidencias
+ search(): Retrona un objeto Match si encuentra una coincidencia
+ split(): Retorna una lista donde se separa la cadena en cada coincidencia
+ sub(): Reemplaza una o más coincidencias por una cadena especificada

Cuenta con metacaracteres, sequencias especiales y conjuntos que terminan
siendo muy útiles a la hora de construir patrones de cadenas.

Referencia: https://www.w3schools.com/python/python_regex.asp

**re.sub en el código**

+ En primera instancia busca reemplazar el link al final del tweet original por un vacío, en este caso borrándolo, utiliza sequencias especiales para indicar el inicio de la ruta del link y su contenido
+ El la segunda instancia busca reemplazar todos los '#' por espacios vacíos 


In [27]:
tweet3 = tweet2.lower()
print("Lowercase tweet: ")
print(tweet3)

Lowercase tweet: 
my beautiful sunflowers on a sunny friday morning off :) sunflowers favourites happy friday off… 


# Unidad 2

**Ejercicio 4.**


$$Ro = \begin{bmatrix} cos \theta & -sin \theta \\ sin \theta & cos \theta \end{bmatrix}$$
$$x = \begin{bmatrix} 2 \\  2 \end{bmatrix}$$

Aplique la siguiente operación
 $$y = x \cdot Ro$$

In [ ]:
# ayuda - No modificar
angle = 100 * (np.pi / 180) # parametro para las funciones 

Salida esperada
```
[[ 1.62231915 -2.31691186]]
```

**Ejercicio 5.**

Aplique la siguientes operaciones a `A`

$$\|\mathrm{A}\|_{F} \equiv \sqrt{\sum_{i=1}^{m} \sum_{j=1}^{n}\left|a_{i j}\right|^{2}}$$


In [ ]:
A = np.array([[2, 2],
              [2, 2]])

Salida esperada
```
4
```

***

Ejecute las siguientes lineas de código para los ejercicios 6-7
```python
from sklearn.datasets import load_wine
import pandas as pd

df = pd.DataFrame(load_wine().data, columns= load_wine().feature_names)
df["target"] = load_wine().target
```

In [35]:
from sklearn.datasets import load_wine
import pandas as pd

df = pd.DataFrame(load_wine().data, columns= load_wine().feature_names)
df["target"] = load_wine().target

**Ejercicio 6.**

¿Cual es la clase mayoritaria ("target") del dataset?

**Ejercicio 7.**

Para la variable `flavanoids` que variables presentan una correlación linear positiva (mayor o igual a 0.6) y negativa (menor o igual a -0.5)

**Ejercicio 8**

En las siguientes celdas se cargará en memoria una representación de las palabras en su forma vectorial. Ejecute e interprete las siguientes celdas

In [52]:
import pickle

word_embeddings = pickle.load(open("word_embeddings_subset.p", "rb"))
len(word_embeddings) # 243 palabras con su representación vectorial

243

In [53]:
print("dimension: {}".format(word_embeddings['Spain'].shape[0])) # cada palabra cuenta con una dimensión de 300 valores

dimension: 300


In [55]:
word_embeddings['Spain'][:10] # los 10 primeros valores de la palabra Spain

array([ 0.03149414, -0.18261719,  0.16308594,  0.12451172, -0.01855469,
       -0.17382812, -0.12109375, -0.1328125 , -0.125     ,  0.24511719],
      dtype=float32)

*Calcule la similitud coseno de las palabras king y queen*

$\cos (\theta)=\frac{\mathbf{A} \cdot \mathbf{B}}{\|\mathbf{A}\|\|\mathbf{B}\|}=\frac{\sum_{i=1}^{n} A_{i} B_{i}}{\sqrt{\sum_{i=1}^{n} A_{i}^{2}} \sqrt{\sum_{i=1}^{n} B_{i}^{2}}}\tag{1}$

<img src="fig.png">

+ Los numeros entre 0 y 1 indican similitud
+ Los numeros entre -1-0 indican una disimilitud 

¿Cual es su interpretación del indicador en las palabras seleccionadas?

## Unidad 3

Entrene un modelo basado en el algortimo de Naive Bayes a partir de la ejecución de las siguientes lineas de código.

```python
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB

X, y = load_wine(return_X_y=True)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)
gnb = GaussianNB().fit(X_train, y_train)
y_pred = gnb.predict(X_test)

from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))
```

In [33]:
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB

X, y = load_wine(return_X_y=True)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)
gnb = GaussianNB().fit(X_train, y_train)
y_pred = gnb.predict(X_test)

from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.88      1.00      0.93        14
           1       1.00      0.81      0.90        16
           2       0.86      1.00      0.92         6

    accuracy                           0.92        36
   macro avg       0.91      0.94      0.92        36
weighted avg       0.93      0.92      0.92        36



**Ejercicio 9.**
+ Investigue e interprete los resultados de la precision

La precisión del modelo hace referencia a ¿qué proporción de todas las predicciones de pertenencia a la clase que hizo el clasificador fueron correctas? Es el resultado de los **Verdaderos Positivos/(Verdaderos Positivos+Falsos Positivos)**
En otras palabras es la estadística que nos dice la calidad del modelo en términos de predecir correctamente para esa clase.

Para el caso que tenemos, podemos observar que para la clase 0 la precisión es de 0.88 es decir que de los 14 acertó en aproximadamente 12. Para la clase 1 es 1.00, que significa que acertó para las 16 ocurrencias. Finalmente para la clase 2 fue de 0.86, que de 6 ocurrencias vendría siendo aproximadamente 5.


**Ejercicio 10.**
+ Investigue e interprete los resultados de recall

El recall del modelo o sensibilidad hace referencia a ¿qué proporción de todos los ejemplos que pertenecían a la clase en la realidad, identificó el clasificador? Es el resultado de **Verdaderos Positivos/(Verdaderos Positivos+Falsos Negativos)**
En otras palabras es la estadística que nos informa sobre la cantidad que el modelo es capaz de identificar para esa clase.

Para el caso que tenemos podemos observar que para las clases 0 y 1 el recall es 1.00 es decir que fue capaz de identificar el 100% de los datos pertenecientes a la clase, en ninguna ocasión predijo que no pertenecía cuando en realidad si. Para la clase 1 el recall fue de 0.81 que nos indica que aproximadamente 4 datos fueron identificados como no pertenecientes a esta clase cuando en realidad sí.


**Ejercicio 11.**
+ Investigue e interprete los resultados de accuracy

El accuracy o exactitud del modelo hace referencia a el porcentaje de caso que acertó el modelo. Es de las métricas más usadas y favoritas, sin embargo, se debe tener cuidado porque no necesariamente indica qué tan bueno es el modelo.

Es el resultado de **Verdaderos Positivos + Verdaderos Negativos /(Verdaderos Positivos+ Verdaderos Negativos+ Falsos Positivos + Falsos Negativos)**

En nuestro caso, la exactitud es de 0.92, que nos indica que el modelo acierta el 92% de las veces, esto es muy bueno. Sin embargo debemos tener cuidado cuando las clases están desbalanceadas.

Referencia: https://www.iartificial.net/precision-recall-f1-accuracy-en-clasificacion/


## Unidad 4

**Ejercicio 12.**

+ Instale la libreria statsmodels e imprima su versión

Ejecute las siguientes lineas de código

```python
import pandas as pd
from statsmodels.formula.api import ols

df = pd.read_csv("taiwan_real_estate2.csv")
```

+ Ejecute, investigue e interprete la siguiente linea de código.

```python
mdl_price_vs_conv = ols("price_twd_msq ~ n_convenience", data=df).fit()
```

**Ejercicio 13.**

Ejecute, investigue e interprete el resultado del $R-squared$
```python
print(mdl_price_vs_conv.summary())
```

**Ejercicio 14.**

La siguiente linea de código nos permite acceder a los componentes de la regresión lineal simple 
```python
mdl_price_vs_conv.params
```

$ Y = a + b*X$

Ejecute la siguiente linea de código para crear la varible independiente $X$

```python
explanatory_data = pd.DataFrame({'n_convenience': np.arange(0,11)})
```
+ Complete $ Y = a + b*X$ y compare los resultados del método predict de `mdl_price_vs_conv` con $X$

## Unidad 5

Ejecute la siguientes lineas de código 

```python
import numpy as np
import matplotlib.pyplot as plt

rgb = np.array([[[255, 0, 0], [255, 0 ,0], [255, 0, 0]],
              [[0, 255, 0], [0, 255, 0], [0, 255, 0]],
              [[0, 0, 255], [0, 0, 255], [0, 0, 255]]])

plt.imshow(rgb)
plt.show()
```

Ejecute las siguiente lineas de código 

```python
colors = np.array([[[255, 0, 0], [255, 255 ,0], [255, 255, 255]],
              [[255, 0, 255], [0, 255, 0], [0, 255, 255]],
              [[0, 0, 0], [0, 255, 255], [0, 0, 255]]])

plt.imshow(colors)
plt.show()
```

**Ejercicio 15.**

+ Utilice el método where de numpy para cambiar los valores de `colors` de 255 a 50
+ utilice imshow() para mostrar el resultado 

Ejecute las siguientes lineas de código
```python
from matplotlib import image
from matplotlib import pyplot

ima = image.imread('hack')
pyplot.imshow(ima)
print(type(ima))
```

**Ejercicio 16.**

+ Obtenga la dimensión del verde de `ima`
+ Utilice imshow() para mostrar el resultado

**Ejercicio 17.**

+ Investigue y utilice un método en numpy que le permita dar vuelta a `ima` 
+ Utilice imshow() y show() para mostrar el resultado